### PR TITLE
feat: introduce `string_implicit_backslashes` as `escape_implicit_backslashes` replacement

### LINKED ISSUE #7621: `escape_implicit_backslashes` should also unfix escaped backslashed
## Feature request

`escape_implicit_backslashes` is a great rule, but it seems it can only add backslashes but not to remove them if redundant.

I would expect the `'single_quoted': false` option to remove redundant backslashes instead of dooing nothing.

Currently, escaped/non-escaped backslashes in single quoted string are kept in both forms leaving the CS very inconsistent.

Test case:
```diff
-in
+expected fix
-$a = '\\a\\';
+$a = '\a\\';
```
(note the end backslash must be always escaped to prevent escaping the end single quote)

### PR DESCRIPTION
**Pull request by mvorisek**
fix #7621

if backslashes are disabled to be explicitly escaped, they must be unescaped

### CODE REVIEW

**Comment by mvorisek** [comment]
PR is done. Tested with
```
'escape_implicit_backslashes' => ['double_quoted' => false, 'heredoc_syntax' => false, 'single_quoted' => true],
```
configuration on the whole project, ie. the inverse configuration. Once removed and fixer is run again, all code returned to exactly the same version as before.

File: tests/Fixer/Casing/ClassReferenceNameCasingFixerTest.php
Comment (mvorisek) [inline tests/Fixer/Casing/ClassReferenceNameCasingFixerTest.php:200]: nice example of CS inconsistency this fix fixes

**Comment by mvorisek** [comment]
@Wirone can you please review this PR as well?

**Comment by mvorisek** [comment]
Hi @Wirone, can you please review this PR?

**Comment by Wirone** [comment]
I hope to do some reviews in the late evening 🤞.

**Comment by mvorisek** [comment]
@Wirone is now a good late evening?

**Comment by julienfalque** [comment]
Shouldn't this new behavior be behind an option?

**Comment by Wirone** [comment]
> @Wirone is now a good late evening?

Hah, with 3 kids you can plan and end with totally different reality 😅. Today I am going on business trip for 2 days, many hours in the train, so I again plan to do some reviews 😉.

**Comment by mvorisek** [comment]
> Shouldn't this new behavior be behind an option?

What about `unescape` option?

It is easy to add, but I am a little worried to whom it will be helpful in reality.

**Comment by Wirone** [comment]
To be honest, I don't know what to do with this PR 🤷‍♂️. In general, fixers are named after the state desired _after_ fix is applied, in this case: all implicit backslashes should be escaped. It's clear that we can't change the current behaviour just like that, it must be introduced behind configuration option. But even if we do this, the idea behind fixer _under this particular name_ would be broken, so in order to do it properly we should introduce new fixer, e.g. `implicit_backslashes` and modify `escape_implicit_backslashes` to be only proxy fixer to the new one, with proper config. I believe the options `double_quoted`, `single_quoted` and `heredoc_syntax` should be a three-state: `?bool` or dictionary-like `escape|unescape|ignore`, in order to achieve both strategies and to be able to leave code as-is.

But it's only my perspective, I believe we need @keradus' opinion.

**Comment by mvorisek** [comment]
I like the 3-state concept and new fixer name - is there any "proxy migration fixer PR" I can as a reference?

**Comment by Wirone** [comment]
You can take a look at #7053, but I would recommend waiting for @keradus' opinion, as I won't merge it anyway without his decision about it 🙂.

**Comment by julienfalque** [comment]
I also like [@Wirone's proposal](#issuecomment-1881582832) and agree the current name of the fixer might not be very accurate with this new behavior.

I do think unescaping should be the default behavior but that might be an unexpected change for users, hence the option. We might reconsider this when we'll prepare a new major version.

File: src/Fixer/StringNotation/EscapeImplicitBackslashesFixer.php
Comment (mvorisek) [inline src/Fixer/StringNotation/EscapeImplicitBackslashesFixer.php:111]: @Wirone how to proxy cnf, ie. how to setup the new fixer using the old/current configuration? The `$this->configuration` is null.

Comment (Wirone) [reply src/Fixer/StringNotation/EscapeImplicitBackslashesFixer.php]: In the old fixer (the one that becomes a proxy) you just configure a new fixer with values that represent old behaviour. Please take a look a other fixers that extend `AbstractProxyFixer` and at the same time implement `ConfigurableFixerInterface`, for example [this](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/860da6f34e2f9965aa922d5e0572b3aab4c0baa9/src/Fixer/Phpdoc/PhpdocNoAliasTagFixer.php) 🙂.

PS. You also need to implement `DeprecatedFixerInterface`

Comment (mvorisek) [reply src/Fixer/StringNotation/EscapeImplicitBackslashesFixer.php]: Thank you a lot for your support!

File: src/Fixer/StringNotation/StringImplicitBackslashesFixer.php
Comment (Wirone) [inline src/Fixer/StringNotation/StringImplicitBackslashesFixer.php:51]: Quick feedback, I will look more later (maybe I can find some time this evening).

```suggestion
            'Handles implicit backslashes in strings and heredocs. Depending on the chosen strategy, it can escape implicit backslashes to ease the understanding of which are special chars interpreted by PHP and which not (`escape`), or it can remove these additional backslashes if you find them superfluous (`unescape`). You can also leave them as-is using `ignore` strategy.',
```

File: tests/Fixer/StringNotation/StringImplicitBackslashesFixerTest.php
Comment (Wirone) [inline tests/Fixer/StringNotation/StringImplicitBackslashesFixerTest.php:664]: Can we _not_ mix code snippets that should be ignored with ones that should be fixed? It's hard to grasp, I had to copy input/expected into Kdiff3 to see what's the difference between them:

image

PS. Info for reviewers: cases above are more or less imported from `EscapeImplicitBackslashesFixerTest` (technical changes only, like heredoc closing markers), this case and cases below are new.

File: src/Fixer/StringNotation/StringImplicitBackslashesFixer.php
Comment (Wirone) [inline src/Fixer/StringNotation/StringImplicitBackslashesFixer.php:1]: FYI: diff between new fixer and `EscapeImplicitBackslashesFixerTest` (not counting supported options, docs etc):

image

**Review by Wirone — CHANGES_REQUESTED** [review]
I would like to see small improvement in tests, other than that - looks good.

File: tests/Fixer/StringNotation/StringImplicitBackslashesFixerTest.php
Comment (mvorisek) [reply tests/Fixer/StringNotation/StringImplicitBackslashesFixerTest.php:664]: Here we test 1..~8 backslashes, the test should stay as is, as in different styles (single quoted, double quoted, heredoc) it must behave differently with different escapes - but the input must be 1..~backslashes and +-same for all styles.

Comment (Wirone) [reply tests/Fixer/StringNotation/StringImplicitBackslashesFixerTest.php:664]: I would argue that it would be better if each test case was split into 2 cases - one with lines that should remain untouched (and with `null` input, so Fixer's test case know that expected code is the same as input), second with lines that should be changed. Right now it's hard to see what is the purpose of the test case, because large part of input code remains untouched. Cognitive load would be much lower, because looking at the test case without input+expected I explicitly know nothing changes in the snippet. Looking at your test cases I don't know what is expected to change and what should be untouched. Each test case's snippet should be a minimal reproducer, without noise.

Comment (mvorisek) [reply tests/Fixer/StringNotation/StringImplicitBackslashesFixerTest.php:664]: I am the author who developed the fixer and these tests helped me as is. When the fixer behaves not as expected, IMHO these tests are much more easier to understand as is. If you want to have them in a different way, feel free to do so, but I do not support this change and once the PR is merged, these tests are not to be edit soon - user can always diff them.

Comment (Wirone) [reply tests/Fixer/StringNotation/StringImplicitBackslashesFixerTest.php:664]: It does not matter who's the author in this case - fixer's code and its test cases must be understandable for all the maintainers, because at any point we can't demand you (or any other contributor) to provide fixes 🙂. For me these cases are not friendly and I don't want to need to copy and diff it externally to understand them. I believe that most of the tests merged during my tenure here had this convention because it's simply more readable (cases names starting with `ignore` + explicit untouched input).

Of course you can stick with your opinion, but then I can stick with my review state 😅. Especially that I don't have much time recently and I can't implement all my expectations on my own, because then I don't have time for my own ideas.

Comment (mvorisek) [reply tests/Fixer/StringNotation/StringImplicitBackslashesFixerTest.php:664]:
```
                \
                \\
                \\\
                \\\\
                \\\\\
                \\\\\\
                \\\\\\\
```

Is a really consistent input noone can break, if we start splitting it, it will become really much harder to read, develop, and maintain, so please, reconsider this as answered. Thank you.

Comment (Wirone) [reply tests/Fixer/StringNotation/StringImplicitBackslashesFixerTest.php:664]: > Is a really consistent input noone can break, if we start splitting it, it will become really much harder to read, develop, and maintain, so please, reconsider this as answered. Thank you.

I have totally different opinion about it, for me it's unreadable in current form because it's nearly impossible to grasp expected change by looking at the input/expected strings. @keradus @julienfalque @kubawerlos please share your thought about it is [this](https://github.com/mvorisek/PHP-CS-Fixer/blob/norm_escape_7621/tests/Fixer/StringNotation/StringImplicitBackslashesFixerTest.php#L664-L721) readable for you?

Comment (mvorisek) [reply tests/Fixer/StringNotation/StringImplicitBackslashesFixerTest.php]: done

Comment (Wirone) [reply tests/Fixer/StringNotation/StringImplicitBackslashesFixerTest.php]: I still see some mixed test cases, where part of the snippet is fixed and part not, but at least these are not mixed that much. I can live with it, however in the future _please_ separate these 🙂. It's much more readable when the snippet contains only bare minimum required to show the case and focus is on _one_ behavior (either on _fix_ or on _ignoring_). Yields' keys also are very helpful to understand the purpose of the test case.

**Review by Wirone — APPROVED** [review]

**Comment by Wirone** [comment]
Thank you @mvorisek 🍻!

**Pull request merged**
